File: ts/project/.sdk/tm/zig/feature/telemetry.zig
// Distributed-tracing telemetry (mirrors go feature/telemetry_feature.go /
// rust feature/telemetry.rs). Opens a span per operation (PrePoint),
// propagates trace context to the server as W3C `traceparent` plus
// `X-Trace-Id` / `X-Span-Id` headers (PreRequest), and closes the span on
// completion (PreDone) or failure (PreUnexpected). Each span closes exactly
// once (the per-context marker in ctx.out is consumed on close). Finished
// spans accumulate on the feature; an `exporter` callback, when provided, is
// invoked with each finished span. Trace/span id generation (`idgen`) and the
// clock (`now`) are injectable for deterministic tests.

const std = @import("std");
const h = @import("../core/helpers.zig");
const types = @import("../core/types.zig");
const sup = @import("support.zig");

const Value = h.Value;
const Context = types.Context;
const Feature = types.Feature;
const OutVal = types.OutVal;

const TELEMETRY_SPAN_KEY: []const u8 = "telemetry_span";

pub const TelemetryFeature = struct {
    name: []const u8 = "telemetry",
    active: bool = true,
    add_opts: Value = .{ .null = {} },
    options: Value = .{ .null = {} },
    seq: i64 = 0,

    // Activity tracking (mirrors the ts client._telemetry record).
    spans: std.ArrayList(Value),
    active_spans: i64 = 0,

    pub fn make() Feature {
        const self = h.A().create(TelemetryFeature) catch unreachable;
        self.* = .{ .spans = std.ArrayList(Value).init(h.A()) };
        return .{ .ptr = @ptrCast(self), .vtable = &vtable };
    }

    fn self_of(p: *anyopaque) *TelemetryFeature {
        return @ptrCast(@alignCast(p));
    }

    fn id(self: *TelemetryFeature, kind: []const u8) []const u8 {
        const idgen = h.getp(self.options, "idgen");
        if (idgen == .function) {
            const r = h.call_vfn(idgen, h.vstr(kind));
            if (r == .string) return r.string;
        }
        // Deterministic-ish sequential id; unique within a client instance.
        self.seq += 1;
        var buf = std.ArrayList(u8).init(h.A());
        buf.writer().print("{x:0>4}", .{@as(u64, @intCast(self.seq))}) catch {};
        while (buf.items.len < 16) buf.append('0') catch {};
        const prefix: []const u8 = if (std.mem.eql(u8, kind, "trace")) "t" else "s";
        return std.fmt.allocPrint(h.A(), "{s}{s}", .{ prefix, buf.items }) catch "";
    }

    fn close(self: *TelemetryFeature, ctx: *Context, ok: bool) void {
        // Close once per operation; a PreDone followed by a pipeline failure
        // (non-2xx) fires PreUnexpected too, which then finds no open span.
        const taken = ctx.out.fetchRemove(TELEMETRY_SPAN_KEY) orelse return;
        const span: Value = switch (taken.value) {
            .val => |v| if (v == .object) v else return,
            else => return,
        };

        const end = sup.fopt_now(self.options);
        const start = h.get_i64(span, "start") orelse 0;
        const dur = @max(end - start, 0);
        h.setp(span, "end", h.vnum(end));
        h.setp(span, "durationMs", h.vnum(dur));
        h.setp(span, "ok", h.vbool(ok));

        self.active_spans -= 1;
        self.spans.append(span) catch {};

        const exporter = h.getp(self.options, "exporter");
        if (exporter == .function) {
            _ = h.call_vfn(exporter, span);
        }
    }

    fn pre_point(self: *TelemetryFeature, ctx: *Context) void {
        if (!self.active) return;

        const entity = ctx.op.entity;
        const opname = ctx.op.name;

        const span = h.omap();
        h.setp(span, "traceId", h.vstr(self.id("trace")));
        h.setp(span, "spanId", h.vstr(self.id("span")));
        h.setp(span, "name", h.vstr(std.fmt.allocPrint(h.A(), "{s}.{s}", .{ entity, opname }) catch ""));
        h.setp(span, "start", h.vnum(sup.fopt_now(self.options)));
        ctx.out_set(TELEMETRY_SPAN_KEY, OutVal{ .val = span });
        self.active_spans += 1;
    }

    fn pre_request(self: *TelemetryFeature, ctx: *Context) void {
        if (!self.active) return;

        const span = ctx.out_val(TELEMETRY_SPAN_KEY);
        if (span != .object) return;
        const sp = ctx.spec orelse return;

        const headers: Value = if (sp.headers == .object) sp.headers else blk: {
            const nh = h.omap();
            sp.headers = nh;
            break :blk nh;
        };

        const hmap = sup.fopt_map(self.options, "headers");
        const trace_id = h.get_str(span, "traceId") orelse "";
        const span_id = h.get_str(span, "spanId") orelse "";
        h.setp(headers, sup.fopt_str(hmap, "trace", "X-Trace-Id"), h.vstr(trace_id));
        h.setp(headers, sup.fopt_str(hmap, "span", "X-Span-Id"), h.vstr(span_id));
        h.setp(headers, sup.fopt_str(hmap, "parent", "traceparent"), h.vstr(std.fmt.allocPrint(h.A(), "00-{s}-{s}-01", .{ trace_id, span_id }) catch ""));
    }

    fn pre_done(self: *TelemetryFeature, ctx: *Context) void {
        const ok = if (ctx.result) |r| (r.ok and r.err == null) else false;
        self.close(ctx, ok);
    }

    fn pre_unexpected(self: *TelemetryFeature, ctx: *Context) void {
        self.close(ctx, false);
    }

    fn vname(p: *anyopaque) []const u8 {
        return self_of(p).name;
    }
    fn vactive(p: *anyopaque) bool {
        return self_of(p).active;
    }
    fn vaddopts(p: *anyopaque) Value {
        return self_of(p).add_opts;
    }
    fn vinit(p: *anyopaque, ctx: *Context, options: Value) void {
        _ = ctx;
        const self = self_of(p);
        self.options = options;
        self.active = sup.fopt_bool(options, "active", false);
        self.seq = 0;
    }
    fn vdispatch(p: *anyopaque, name: []const u8, ctx: *Context) void {
        const self = self_of(p);
        if (std.mem.eql(u8, name, "PrePoint")) {
            self.pre_point(ctx);
        } else if (std.mem.eql(u8, name, "PreRequest")) {
            self.pre_request(ctx);
        } else if (std.mem.eql(u8, name, "PreDone")) {
            self.pre_done(ctx);
        } else if (std.mem.eql(u8, name, "PreUnexpected")) {
            self.pre_unexpected(ctx);
        }
    }

    const vtable = Feature.VTable{
        .name = vname,
        .active = vactive,
        .add_options = vaddopts,
        .init = vinit,
        .dispatch = vdispatch,
    };
};

File: ts/project/.sdk/tm/rust/feature/telemetry.rs
// Distributed-tracing telemetry (mirrors go feature/telemetry_feature.go).
// Opens a span per operation (PrePoint), propagates trace context to the
// server as W3C `traceparent` plus `X-Trace-Id` / `X-Span-Id` headers
// (PreRequest), and closes the span on completion (PreDone) or failure
// (PreUnexpected). Each span closes exactly once (the per-context marker
// in ctx.out is consumed on close). Finished spans accumulate on the
// feature; an `exporter` callback, when provided, is invoked with each
// finished span. Trace/span id generation (`idgen`) and the clock (`now`)
// are injectable for deterministic tests.

use std::rc::Rc;

use crate::core::context::Context;
use crate::core::helpers::{call_vfn, get_i64, get_str, getp, setp};
use crate::core::types::{Feature, OutVal};
use crate::feature::support::*;
use crate::utility::voxgigstruct::Value;

const TELEMETRY_SPAN_KEY: &str = "telemetry_span";

pub struct TelemetryFeature {
    pub name: String,
    pub active: bool,
    pub add_opts: Option<Value>,
    options: Value,
    seq: i64,

    // Activity tracking (mirrors the ts client._telemetry record).
    pub spans: Vec<Value>,
    pub active_spans: i64,
}

impl TelemetryFeature {
    pub fn new() -> TelemetryFeature {
        TelemetryFeature {
            name: "telemetry".to_string(),
            active: true,
            add_opts: None,
            options: Value::Noval,
            seq: 0,
            spans: Vec::new(),
            active_spans: 0,
        }
    }

    fn id(&mut self, kind: &str) -> String {
        let idgen = getp(&self.options, "idgen");
        if let Value::Func(_) = idgen {
            if let Value::Str(s) = call_vfn(&idgen, &Value::str(kind)) {
                return s;
            }
        }
        // Deterministic-ish sequential id; unique within a client instance.
        self.seq += 1;
        let mut n = format!("{:04x}", self.seq);
        let prefix = if kind == "trace" { "t" } else { "s" };
        while n.len() < 16 {
            n.push('0');
        }
        format!("{}{}", prefix, n)
    }

    fn close(&mut self, ctx: &Rc<Context>, ok: bool) {
        // Close once per operation; a PreDone followed by a pipeline failure
        // (non-2xx) fires PreUnexpected too, which then finds no open span.
        let span = match ctx.out_take(TELEMETRY_SPAN_KEY) {
            Some(OutVal::Val(v)) if matches!(v, Value::Map(_)) => v,
            _ => return,
        };

        let end = fopt_now(&self.options)();
        let start = get_i64(&span, "start").unwrap_or(0);
        let dur = (end - start).max(0);
        setp(&span, "end", Value::Num(end as f64));
        setp(&span, "durationMs", Value::Num(dur as f64));
        setp(&span, "ok", Value::Bool(ok));

        self.active_spans -= 1;
        self.spans.push(span.clone());

        let exporter = getp(&self.options, "exporter");
        if let Value::Func(_) = exporter {
            call_vfn(&exporter, &span);
        }
    }
}

impl Feature for TelemetryFeature {
    fn name(&self) -> String {
        self.name.clone()
    }
    fn active(&self) -> bool {
        self.active
    }
    fn add_options(&self) -> Option<Value> {
        self.add_opts.clone()
    }

    fn init(&mut self, _ctx: &Rc<Context>, options: &Value) {
        self.options = options.clone();
        self.active = fopt_bool(options, "active", false);
        self.seq = 0;
    }

    fn pre_point(&mut self, ctx: &Rc<Context>) {
        if !self.active {
            return;
        }

        let (entity, opname) = {
            let op = ctx.op.borrow();
            (op.entity.clone(), op.name.clone())
        };

        let span = Value::empty_map();
        setp(&span, "traceId", Value::str(self.id("trace")));
        setp(&span, "spanId", Value::str(self.id("span")));
        setp(&span, "name", Value::str(format!("{}.{}", entity, opname)));
        setp(
            &span,
            "start",
            Value::Num(fopt_now(&self.options)() as f64),
        );
        ctx.out_set(TELEMETRY_SPAN_KEY, OutVal::Val(span));
        self.active_spans += 1;
    }

    fn pre_request(&mut self, ctx: &Rc<Context>) {
        if !self.active {
            return;
        }

        let span = ctx.out_val(TELEMETRY_SPAN_KEY);
        let spec = ctx.spec.borrow().clone();
        let spec = match (matches!(span, Value::Map(_)), spec) {
            (true, Some(s)) => s,
            _ => return,
        };

        let headers = {
            let h = spec.borrow().headers.clone();
            if matches!(h, Value::Map(_)) {
                h
            } else {
                let nh = Value::empty_map();
                spec.borrow_mut().headers = nh.clone();
                nh
            }
        };

        let h = fopt_map(&self.options, "headers");
        let trace_id = get_str(&span, "traceId").unwrap_or_default();
        let span_id = get_str(&span, "spanId").unwrap_or_default();
        setp(
            &headers,
            &fopt_str(&h, "trace", "X-Trace-Id"),
            Value::str(trace_id.clone()),
        );
        setp(
            &headers,
            &fopt_str(&h, "span", "X-Span-Id"),
            Value::str(span_id.clone()),
        );
        setp(
            &headers,
            &fopt_str(&h, "parent", "traceparent"),
            Value::str(format!("00-{}-{}-01", trace_id, span_id)),
        );
    }

    fn pre_done(&mut self, ctx: &Rc<Context>) {
        let ok = match ctx.result.borrow().clone() {
            Some(r) => {
                let r = r.borrow();
                r.ok && r.err.is_none()
            }
            None => false,
        };
        self.close(ctx, ok);
    }

    fn pre_unexpected(&mut self, ctx: &Rc<Context>) {
        self.close(ctx, false);
    }
}
